### PR TITLE
[JENKINS-52409] Add PARENT_ITEM_NAME env var

### PR DESCRIPTION
**Pull request by rabadin**
Symmetric to the addition of the JOB_BASE_NAME env var in
4bab3630efcc2fa, this change adds the PARENT_ITEM_NAME
env var.

Folder names usually represent organisational structure that
is useful to have exposed in builds.

### CODE REVIEW

**Comment by rabadin** [comment]
@fbelzunc This change is a "complement" to the change you did in 4bab3630efcc2fa3, maybe you will want to have a look.

**Comment by daniel-beck** [comment]
I wonder whether there's a better name for this, as it's ambiguous: My first assumption here was that it provides a file system path. The symmetry matching `basename`/`dirname` makes sense OTOH.

**Comment by rabadin** [comment]
@daniel-beck I'm open to suggestions for a better name; but, as you pointed out, the well-understood `basename` / `dirname` concepts make `JOB_DIR_NAME` a logical name after `JOB_BASE_NAME` was introduced.

**Comment by daniel-beck** [comment]
FWIW I've only ever seen these called *Folder* in Jenkins, and the plugin doing it most of the time is the Folder Plugin. Perhaps that would be a way out?

**Comment by rabadin** [comment]
@daniel-beck Sure, we could rename the variable `JOB_FOLDER_NAME`.

**Comment by rabadin** [comment]
@daniel-beck I've updated the PR.

File: core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.properties
Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.properties]: Terminology inconsistent after rename. Perhaps "Full name of the parent folder of the current job"?

Comment (rabadin) [reply core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.properties]: Good point; I've changed it to 'Full name of the parent folder of this build, such as "bar" for "bar/foo".' to be consistent with the other messages.

Comment (daniel-beck) [reply core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.properties]: We distinguish between projects and builds, and the new version of this message does not. See the message for JOB_BASE_NAME: "Name of the project of this build".

(Builds can be hierarchical (Maven and Matrix job types do this), but it's a parallel hierarchy to corresponding jobs.)

File: core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.properties
Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.properties]: Should explicitly mention what the folder name is in case of a top level job, and a three level nested hierarchy.

Comment (rabadin) [reply core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.properties]: Sure: done.

**Comment by rabadin** [comment]
@daniel-beck I've reworked the wording of the commit message and the PR title to align with the changes you suggested during review.  I'm unclear why the unit-tests are failing thought, I don't really understand how my change can directly cause the test `processProperlyKilledUnix` to fail; would you mind taking a look (this might be a known spurious failure)?

**Review by oleg-nenashev — COMMENTED** [review]
Maybe just "JOB_PARENT_FULL_NAME"? Why, not all parents are actually folders (e.g. in Promoted builds and Matrix Project)

Apart from the variable name, it looks good. Test failure is unrelated?

**Comment by rabadin** [comment]
@oleg-nenashev Sure, works for me if you think this is a better name.  I've made the change.  CI failure looks totally unrelated to me.  If this change looks good, would you be okay to approve it and get it landed?

**Review by oleg-nenashev — APPROVED** [review]
Looks good to me in the current state.
I will merge it towards next weekly if there is no negative feedback from @daniel-beck or other reviewers

**Comment by rabadin** [comment]
Cool, thanks a lot @oleg-nenashev !

**Comment by daniel-beck** [comment]
Would like to take another look at this.

Can't help the impression that `JOB_` in this is redundant (`BUILD_URL` doesn't need it). For consistency, `PARENT_NAME` / `PARENT_URL` (or `FOLDER_*`) would make sense.

**Comment by rabadin** [comment]
@daniel-beck @oleg-nenashev I wonder if we're not overengineering this a tad but at the same time I appreciate that you guys are trying to make sure the naming of a variable that's here to stay is consistent.
Now, to take into account @oleg-nenashev 's remarks and your last remark @daniel-beck I believe `PARENT_FULL_NAME` might be a good option.  It's basically what @daniel-beck suggested plus `FULL` which I believe is important because it denotes clearly that we're talking about the full path, starting from the top level, not just the name of the parent.

@oleg-nenashev @daniel-beck Would you guys agree with the last version, i.e. `PARENT_FULL_NAME`?

**Review by oleg-nenashev — APPROVED** [review]
@rabadin Thanks a lot for your patience, and sorry if the review process causes some frustration. In this case this is "once merged - stays forever" thing, so that's why there is such discussion. I am fine with the current name though I like the previous one more. Anyway, no need to revert it again. I am happy to make such changes for you if we decide to change the name again

@daniel-beck I would still prefer "JOB_" prefix though, because some build types may also have parents (e.g. Matrix Build). And such item may also have a full name. In the current state the name may be a bit confusing for those users who do not read documentation.

**Comment by rabadin** [comment]
@oleg-nenashev As I said above, I totally understand that we want to be sure this is named appropriately: once this is checked in and people start relying on it, there is no going back. Once you and @daniel-beck agree on the proper name for this, I'm more than happy to make the change.  Please let me know what you decide!

**Comment by oleg-nenashev** [comment]
@daniel-beck gentle ping

**Comment by daniel-beck** [comment]
On my to-do list for next week. My apologies for the delay.

(FWIW the current iteration adding `FULL` to the name seems confusingly inconsistent, as `JOB_NAME` _is_ the job's full name, so we needed `JOB_BASE_NAME` for the not-full name. That's as far as I got. Sorry.)

**Comment by rabadin** [comment]
@daniel-beck Makes sense, I changed it back to `PARENT_NAME`.

**Comment by daniel-beck** [comment]
> On my to-do list for next week.

Didn't find the time and on the road the next few days. Sorry again.

**Comment by daniel-beck** [comment]
Two possible options:

* `PARENT_NAME` as it currently is. Short and to the point.
* `PARENT_ITEM_NAME` if we want to be clear that in the case of hierarchical builds, it's the parent item (job or other item, typically an `ItemGroup`), not the build's parent. `PARENT_BUILD_*` would be the corresponding name for variables in that space then.

I have no strong preference for either. @oleg-nenashev might though.

**Comment by oleg-nenashev** [comment]
Let's do the `PARENT_ITEM_NAME` then.
@rabadin I am happy to do renames in this PR, so you can leave it to me

**Comment by rabadin** [comment]
@oleg-nenashev I just did the rename.

**Review by oleg-nenashev — APPROVED** [review]

**Comment by daniel-beck** [comment]
Tests are failing but unclear why. Looks like an unrelated flake but it's too consistent for that.

**Comment by oleg-nenashev** [comment]
I am trying to reproduce it locally

**Comment by oleg-nenashev** [comment]
I am unable to reproduce it locally.
@daniel-beck my proposal is to integrate it early next week once we finish traveling. So that we can react if something goes wrong

**Comment by rabadin** [comment]
@oleg-nenashev Are you okay to land this now?

**Comment by oleg-nenashev** [comment]
Got the ping, will re-review why the tests fail. Still no idea so far

**Comment by daniel-beck** [comment]
Tests consistently fail here but we don't know why. It seems related to this change though.

**Comment by batmat** [comment]
FTR in case the CI history gets cleaned up, here are the tests consistently failing
```
Linux-jdk11 / Linux Publishing / hudson.util.ProcessTreeKillerTest.processProperlyKilledUnix
Linux-jdk8 / Linux Publishing / hudson.util.ProcessTreeKillerTest.processProperlyKilledUnix
```

**Review by daniel-beck — CHANGES_REQUESTED** [review]
The reason the test fails is that `Launcher#launch` &rarr; `Launcher#inherit` &rarr; `EnvVars#overrideExpandingAll` &rarr; `EnvVars#override(String, String)`, which drops empty values from the list:

https://github.com/jenkinsci/jenkins/blob/e460e6c40b0663165aa81f4420438480a264f1ea/core/src/main/java/hudson/EnvVars.java#L147

This indicates that the behavior documented in https://github.com/jenkinsci/jenkins/pull/3539/files#diff-20ac3a9aa8489ad7a3eaefa8b88cbd08R6 in response to https://github.com/jenkinsci/jenkins/pull/3539#discussion_r200806364 was not confirmed interactively to be true.

---

`getCharacteristicEnvVars` exists to allow Jenkins to kill processes launched during the build that would outlive the build duration. A sometimes empty value does not help accomplish this.

That we've started convenient variables there with `JOB_BASE_NAME` has already been a mistake. This and `PARENT_ITEM_NAME` should be put elsewhere. Just add an `EnvironmentContributor` or something that adds these utility values.

---

The question beyond _where_ to add these variables is whether we want them always defined, or undefined values would indicate a top level project. This could be a problem in scripts intended to be invoked from different Jenkins releases, as a top level project would be indistinguishable from a folder project on an older Jenkins.

`EnvVars#override(String, String)` could be changed to note drop empty values, but that seems like a more complicated change.

I'm not sure what the best way forward is here. Accepting undefined variables in the top level project case seems bad. Requiring a change to `EnvVars` is complicated, and the history of this PR does not fill me with confidence about this approach.

**Comment by daniel-beck** [comment]
Given this has never passed existing tests, and in my recent review I identified substantial problems with the implementation, I'm closing this for general housekeeping.

This isn't a rejection of the overall proposal, but I don't see this specific implementation happening due to the problem in the previous comment.

I recommend you address review comments, add tests confirming the expected behavior for top level project and projects in folders, and rebase the entire change on top of a recent master. I would then not expect further notable obstacles for this change.

**Pull request closed without merging**
